### PR TITLE
py/objstr.c: Enforce unicode-safety when creating str from vstr.

### LINKED ISSUE #9044: various ways to read() don't check for valid utf-8, weirdness ensues
[Initial comment heavily revised because the original was unclear. Apologies if this breaks any of the context of the replies below]

Over in https://github.com/adafruit/circuitpython/issues/6752 a user reported that various ways to read from files didn't actually validate that the content was utf-8. Instead, it blithely stores the data in a `str`, which in Micropython must always be a valid utf-8 encoded chunk of memory.

This leads to weird behavior, such as when the file with the byte content `\xf7` is read:
```python
>>> with open("notutf8.txt", "rt", -1, "utf-8") as f: content = f.read()
... 
>>> f
<io.TextIOWrapper -1>
>>> content
'\x07'
>>> content[0]
'\x07\x00\x00\x00'
>>> 
```

The behavior is the same no matter what coding is specified, or if no coding is explicitly specified. As discussed below, it does not appear to be explicit anywhere that the encoding of a text file is utf-8. It is implicit in the way that reading a file with UTF-8 content works, and reading a file with non-UTF8 content yields up dragons.

I prepared a possible fix relative to CircuitPython at https://github.com/adafruit/circuitpython/pull/6754 -- is there interest in changing this behavior in micropython? if so, I can prepare a PR for y'all including hopefully adding tests.

basically, each code path in file reading will now check that the content is valid utf-8 in the case that a str, rather than a bytes, is being produced.

### PR DESCRIPTION
**Pull request by jimmo**
Fixes #9044 and https://github.com/adafruit/circuitpython/issues/6752 (@jepler @RetiredWizard).

Rather than just ensuring that streams are unicode safe, this ensures that all usages of mp_obj_str_from_vstr are safe, as there are other ways to trigger this bug (see https://github.com/micropython/micropython/issues/9044#issuecomment-1218053443).

By changing the `*_from_vstr` API slightly, it's possible to do this in a way that overall reduces code size and comes in at -104 bytes on PYBV11.

I hope this isn't a huge breaking change for anyone writing user c modules (or for CircuitPython to merge) but it's a very straightforward find-and-replace to fix.

### CODE REVIEW

**Comment by dlech** [comment]
LGTM. Should we add a test, e.g. reading a file or memory stream with bad UTF-8 data?

File: extmod/modure.c [hunk 1]
```diff
@@ -372,7 +372,11 @@ STATIC mp_obj_t re_sub_helper(size_t n_args, const mp_obj_t *args) {
     // Add post-match string
     vstr_add_strn(&vstr_return, subj.begin, subj.end - subj.begin);
 
-    return mp_obj_new_str_from_vstr(mp_obj_get_type(where), &vstr_return);
+    if (mp_obj_get_type(where) == &mp_type_str) {
+        return mp_obj_new_str_from_vstr(&vstr_return);
+    } else {
+        return mp_obj_new_bytes_from_vstr(&vstr_return);
+    }
```
Comment (dpgeorge) [inline extmod/modure.c:379]: Would be interesting to find a test case where doing a re match does raise a unicode error, and then check how CPython handles it.

Also, to reduce code size, maybe a helper like `mp_obj_new_str_type_from_vstr_checked(type, vstr)` could be added?

File: extmod/modujson.c [hunk 2]
```diff
@@ -67,7 +67,7 @@ STATIC mp_obj_t mod_ujson_dump_helper(size_t n_args, const mp_obj_t *pos_args, m
         vstr_t vstr;
         vstr_init_print(&vstr, 8, &print_ext.base);
         mp_obj_print_helper(&print_ext.base, pos_args[0], PRINT_JSON);
-        return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+        return mp_obj_new_str_from_vstr(&vstr);
```
Comment (dpgeorge) [inline extmod/modujson.c:70]: Does this need to validate utf8?  I would think that it's all already been validated as part of the dump process, or at least the dump should never produce invalid utf8.

File: extmod/modujson.c [hunk 3]
```diff
@@ -103,7 +103,7 @@ STATIC mp_obj_t mod_ujson_dumps(mp_obj_t obj) {
     mp_print_t print;
     vstr_init_print(&vstr, 8, &print);
     mp_obj_print_helper(&print, obj, PRINT_JSON);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Comment (dpgeorge) [inline extmod/modujson.c:106]: Does in need validating?

File: extmod/vfs.c [hunk 4]
```diff
@@ -360,7 +360,7 @@ mp_obj_t mp_vfs_getcwd(void) {
     if (!(cwd[0] == '/' && cwd[1] == 0)) {
         vstr_add_str(&vstr, cwd);
     }
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Comment (dpgeorge) [inline extmod/vfs.c:363]: Can a filesystem have invalid chars in it?  How do you get the cwd if it does?  How does CPython handle this?

File: ports/unix/modusocket.c [hunk 5]
```diff
@@ -559,7 +559,7 @@ STATIC mp_obj_t mod_socket_inet_ntop(mp_obj_t family_in, mp_obj_t binaddr_in) {
         mp_raise_OSError(errno);
     }
     vstr.len = strlen(vstr.buf);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Comment (dpgeorge) [inline ports/unix/modusocket.c:562]: Does this need validating?

File: py/modbuiltins.c [hunk 6]
```diff
@@ -467,7 +467,7 @@ STATIC mp_obj_t mp_builtin_repr(mp_obj_t o_in) {
     mp_print_t print;
     vstr_init_print(&vstr, 16, &print);
     mp_obj_print_helper(&print, o_in, PRINT_REPR);
-    return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+    return mp_obj_new_str_from_vstr(&vstr);
```
Comment (dpgeorge) [inline py/modbuiltins.c:470]: Surely the output of PRINT_REPR doesn't need validating?

File: py/parsenum.c [hunk 7]
```diff
@@ -160,7 +160,7 @@ mp_obj_t mp_parse_num_integer(const char *restrict str_, size_t len, int base, m
         mp_printf(&print, "invalid syntax for integer with base %d: ", base);
         mp_str_print_quoted(&print, str_val_start, top - str_val_start, true);
         mp_obj_t exc = mp_obj_new_exception_arg1(&mp_type_ValueError,
-            mp_obj_new_str_from_vstr(&mp_type_str, &vstr));
+            mp_obj_new_str_from_vstr(&vstr));
```
Comment (dpgeorge) [inline py/parsenum.c:163]: Does this need validating?

**Comment by dpgeorge** [comment]
What about calls to `mp_obj_new_str()` and `mp_obj_new_str_of_type()`?  Do they also need to validate?

**Comment by jepler** [comment]
Hi, thanks for making this PR. I look forward to replacing the quick patch in CircuitPython with it down the road, especially if some flash space is freed in the process.

**Comment by jimmo** [comment]
I've added an optimised path `mp_obj_new_str_from_vstr_safe` for the cases outlined above (and one more in persistentcode.c).

> Should we add a test, e.g. reading a file or memory stream with bad UTF-8 data?

Good idea. Done.

> What about calls to `mp_obj_new_str()` and `mp_obj_new_str_of_type()`? Do they also need to validate?

I think the the former should always check (all the uses I could find were small strings or untrusted). I have added a commit to do this.

`mp_obj_new_str_of_type` calls `mp_obj_new_str`.

The other two interesting ones are:

`mp_obj_new_str_copy` -- this is always used on safe strings
`mp_obj_new_str_via_qstr` -- this is also always used on safe strings

I have added/updated the comments for both to state this as a requirement.

---

This now comes in at -72 bytes on PYBV11.

File: py/obj.h [hunk 8]
```diff
@@ -789,11 +789,11 @@ mp_obj_t mp_obj_new_int_from_uint(mp_uint_t value);
 mp_obj_t mp_obj_new_int_from_str_len(const char **str, size_t len, bool neg, unsigned int base);
 mp_obj_t mp_obj_new_int_from_ll(long long val); // this must return a multi-precision integer object (or raise an overflow exception)
 mp_obj_t mp_obj_new_int_from_ull(unsigned long long val); // this must return a multi-precision integer object (or raise an overflow exception)
-mp_obj_t mp_obj_new_str(const char *data, size_t len);
-mp_obj_t mp_obj_new_str_via_qstr(const char *data, size_t len);
-mp_obj_t mp_obj_new_str_from_vstr(vstr_t *vstr);
+mp_obj_t mp_obj_new_str(const char *data, size_t len); // will check utf-8 (raises UnicodeError)
+mp_obj_t mp_obj_new_str_via_qstr(const char *data, size_t len); // must be known to be utf-8 safe
```
Comment (dpgeorge) [inline py/obj.h:793]: "input data must be valid utf-8"

File: extmod/modujson.c [hunk 9]
```diff
@@ -67,7 +67,7 @@ STATIC mp_obj_t mod_ujson_dump_helper(size_t n_args, const mp_obj_t *pos_args, m
         vstr_t vstr;
         vstr_init_print(&vstr, 8, &print_ext.base);
         mp_obj_print_helper(&print_ext.base, pos_args[0], PRINT_JSON);
-        return mp_obj_new_str_from_vstr(&mp_type_str, &vstr);
+        return mp_obj_new_str_from_vstr_safe(&vstr);
```
Comment (dpgeorge) [inline extmod/modujson.c:70]: `mp_obj_new_str_from_utf8_vstr`

**Comment by jimmo** [comment]
Updated.

File: py/objstr.h [hunk 10]
```diff
@@ -88,8 +88,8 @@ mp_obj_t mp_obj_str_make_new(const mp_obj_type_t *type_in, size_t n_args, size_t
 void mp_str_print_json(const mp_print_t *print, const byte *str_data, size_t str_len);
 mp_obj_t mp_obj_str_format(size_t n_args, const mp_obj_t *args, mp_map_t *kwargs);
 mp_obj_t mp_obj_str_split(size_t n_args, const mp_obj_t *args);
-mp_obj_t mp_obj_new_str_copy(const mp_obj_type_t *type, const byte *data, size_t len);
-mp_obj_t mp_obj_new_str_of_type(const mp_obj_type_t *type, const byte *data, size_t len);
+mp_obj_t mp_obj_new_str_copy(const mp_obj_type_t *type, const byte *data, size_t len); // for type=str, must be known to be utf-8 safe
```
Comment (dpgeorge) [inline py/objstr.h:91]: "input data must be valid utf-8"

**Comment by dpgeorge** [comment]
In `mp_obj_str_make_new` there are two calls to `utf8_check`.  I think the second can be removed, and I think the first needs to stay.

**Comment by jimmo** [comment]
> In `mp_obj_str_make_new` there are two calls to `utf8_check`. I think the second can be removed, and I think the first needs to stay.

Yes good catch. Updated.

Now at -80 bytes.

**Comment by dpgeorge** [comment]
Thanks for updating (and I also updated the code comment as per above).  Rebased and merged in 8a0ee5a5c04e83f04d1c62029ac5ba7c74856507 through bd4e45fd68e20af3f965b403e643fa5f71aa1b08

**Comment by dpgeorge** [comment]
I also run the performance benchmarks before/after this PR, on PYBv1.0.  It was basically unchanged (as expected because there aren't any tests for string operations).

**Pull request closed without merging**